server/tserver/src/main/java/org/apache/accumulo/tserver/BusiestTracker.java

```diff
@@ -0,0 +1,142 @@
+ * limitations under the License.
+ */
+package org.apache.accumulo.tserver;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+import org.apache.accumulo.core.data.impl.KeyExtent;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.accumulo.tserver.tablet.Tablet;
+
+/**
+ * <P>
+ * Computes the N tablets that have the highest deltas for a given monotonically increasing counter.
+ * 
+ * <P>
+ * This class was parameterized to facilitate writing unit test. Writing unit test with the Tablet
+ * type would cumbersome.
+ * 
+ * @param <K>
+ *          Represents a key extent
+ * @param <T>
+ *          Represents a Tablet
+ */
+public abstract class BusiestTracker<K extends Comparable<K>,T> {
+
+  private Map<K,Long> lastCounts = Collections.emptyMap();
+  private final int numBusiestTabletsToLog;
+
+  BusiestTracker(int numBusiestTabletsToLog) {
+    this.numBusiestTabletsToLog = numBusiestTabletsToLog;
+  }
+
+  protected abstract K extractKey(T tablet);
+
+  protected abstract long extractCount(T tablet);
+
+  public List<Pair<K,Long>> computeBusiest(Collection<T> tablets) {
+
+    Comparator<Pair<K,Long>> busiestTabletComparator = new Comparator<Pair<K,Long>>() {
+      @Override
+      public int compare(Pair<K,Long> first, Pair<K,Long> second) {
+        int cmp = first.getSecond().compareTo(second.getSecond());
+        if (cmp == 0) {
+          cmp = first.getFirst().compareTo(second.getFirst());
+        }
+        return cmp;
+      }
+    };
+
+    HashMap<K,Long> counts = new HashMap<>();
+
+    PriorityQueue<Pair<K,Long>> busiestTabletsQueue =
+        new PriorityQueue<>(numBusiestTabletsToLog, busiestTabletComparator);
+
+    for (T tablet : tablets) {
+      K extent = extractKey(tablet);
+
+      // only get the count once to ensure consistency in the case of multiple threads
+      long count = extractCount(tablet);
+
+      if (count == 0)
+        continue;
+
+      counts.put(extent, count);
+
+      Long lastCount = lastCounts.get(extent);
+      if (lastCount == null)
+        lastCount = 0L;
+
+      long delta = count - lastCount;
+
+      // handle case where tablet leaves tserver and comes back OR tablet has had no activity
+      if (delta <= 0)
+        continue;
+
+      if (busiestTabletsQueue.size() < numBusiestTabletsToLog
+          || busiestTabletsQueue.peek().getSecond() < delta) {
+        if (busiestTabletsQueue.size() == numBusiestTabletsToLog) {
+          busiestTabletsQueue.remove();
+        }
+
+        busiestTabletsQueue.add(new Pair<K,Long>(extent, delta));
+      }
+    }
+
+    lastCounts = counts;
+
+    ArrayList<Pair<K,Long>> busyList = new ArrayList<>(busiestTabletsQueue);
+    Collections.sort(busyList, busiestTabletComparator.reversed());
```

Comment:
Could bake the comparator in to a `TopN` data structure, so there's no need to reverse it. Could store it reversed from the start. The `Queue` API is limited for this, though, because it only allows peeking at the one end.

Reply:
Comparator when away.